Process: Add documentation for labels, current and proposed (#533)
webmachinelearning/webnn: IssueTriage.md

# Triage Guidance for WebNN Repo

TODO: Consider removing these, or fit these into the documentation below.

> - **enhancement** - unclear if this is about the spec text, the proposed API, etc. Prefer **feature request** etc.

Reviewer: I did some further triage, and the issues with https://github.com/webmachinelearning/webnn/labels/enhancement are only a few. We should probably drop this label entirely.

- Understanding the scope of work remaining on broad efforts (e.g. aligning with best practices, fixing normative issues, etc)
- Identifying areas to contribute.

The working group chairs and spec editors should regularly review bugs and ensure that labels are accurate, and ensure that issues are getting appropriate attention; for example, scheduling discussion of new feature requests, discussion to resolve outstanding questions, and drawing attention to issues that are ready for a contributor to author a PR.

## Types

Every issue should have one of these issue types, and only rarely more than one.

- **bug** - a flaw in the specification that will require a normative fix; for example, an algorithm computes an output incorrectly
- **conventions** - where the spec does not conform to specification best practices from Web IDL, Bikeshed, Infra, etc.
- **use case** - a new use case for the API that should be documented or considered; may spawn other issues
- **process** - a meta issue about how the specification is evolved; for example, a discussion of issue labels
- **testing** - discussion of test coverage

> - **feature request** - _PROPOSED_ - suggestion for an addition to the proposed API

Reviewer: Added https://github.com/webmachinelearning/webnn/labels/feature%20request to test drive this.

Author: I did a pass and labeled issues that looked like requests for pure functionality additions. How does it look to others?

Reviewer 2: > I'll defer to others who know the space better than I do re: "opset" vs. "op set" vs. "operator set" etc.

I've always just said `opset` as a shorthand for `operator set`. 🤷‍♂️

Reviewer 3: `opset` for label sgtm.

Author: Updated proposal w/ `opset`

Reviewer: Renamed to https://github.com/webmachinelearning/webnn/labels/opset

- **interop** - _PROPOSED_ - discussion of differences in behavior across implementations
- **samples** - _PROPOSED_ - suggestions for additional or improved samples

## Spec Impact

These broad categories describe the projected impact on the specification and implementation of an issue.

- **editorial** - spec text could be improved, but does not imply normative behavior changes to the proposed API
- **normative** - _PROPOSED_ - resolving the issue would require behavior changes, likely affecting implementations and tests

> TODO: I'm not convinced of the utility of these; maybe drop them?

Reviewer: One use case for "editorial" is to filter out issues that are https://www.w3.org/2023/Process-20231103/#editorial-change to assess whether the spec satisfies Rec Track transition requirements at various maturity stages.

I think we don't need normative.

# Triage Guidance for WebNN Repo

TODO: Consider removing these, or fit these into the documentation below.

> - **dependencies** - unused?

Reviewer: I deleted this currently unused label.

The fewer the labels the more likely the group will use them (and use them correctly).

> - **tag** - redundant with **tag-needs-resolution** and **tag-tracker** ?

Reviewer: Ditto. This is now gone.

- **enhancement** - unclear if this is about the spec text, the proposed API, etc. Prefer **feature request** etc.
- **help wanted** - unclear what help is needed; use **needs PR** or **question** instead?

## Label Usage

Labels are used for:

- Understanding the status of a specific issue and next steps to resolve it.
- Understanding the scope of work remaining on broad efforts (e.g. aligning with best practices, fixing normative issues, etc)
- Identifying areas to contribute.

The working group chairs and spec editors should regularly review bugs and ensure that labels are accurate, and ensure that issues are getting appropriate attention; for example, scheduling discussion of new feature requests, discussion to resolve outstanding questions, and drawing attention to issues that are ready for a contributor to author a PR.

## Types

Every issue should have one of these issue types, and only rarely more than one.

- **bug** - a flaw in the specification that will require a normative fix; for example, an algorithm computes an output incorrectly

> - **conventions-integration** - where the spec does not conform to specification best practices from Web IDL, Bikeshed, Infra, etc.

Reviewer: ```suggestion
- **conventions** - where the spec does not conform to specification best practices from Web IDL, Bikeshed, Infra, etc.
```
Renamed this. Short and sweet.

- **use case** - a new use case for the API that should be documented or considered; may spawn other issues
- **process** - a meta issue about how the specification is evolved; for example, a discussion of issue labels
- **testing** - discussion of test coverage
- **feature request** - _PROPOSED_ - suggestion for an addition to the proposed API
- **interop** - _PROPOSED_ - discussion of differences in behavior across implementations
- **samples** - _PROPOSED_ - suggestions for additional or improved samples

## Spec Impact

These broad categories describe the projected impact on the specification and implementation of an issue.

> - **Editorial** - spec text could be improved, but does not imply normative behavior changes to the proposed API
> - **Normative** - _PROPOSED_ - resolving the issue would require behavior changes, likely affecting implementations and tests
> 
> TODO: I'm not convinced of the utility of these; maybe drop them?
> 
> TODO: Make these lowercase for consistency?

Reviewer: ```suggestion
- **editorial** - spec text could be improved, but does not imply normative behavior changes to the proposed API
- **normative** - _PROPOSED_ - resolving the issue would require behavior changes, likely affecting implementations and tests

TODO: I'm not convinced of the utility of these; maybe drop them?
```
Made editorial lowercase.

## Issue Scope

Every issue should have at least one of these issue types, but occasionally multiple.

> - **operation set** - discussions about the overall operator coverage of WebNN; examples include alignment with other published operator sets, use cases that require multiple new operators, compatibility with implementations, etc.
> - **operator specific** - _PROPOSED_ - issues regarding the specification of a single operator or small number of operators

Reviewer: ```suggestion
- **operation set** - discussions about the overall operation coverage of WebNN; examples include alignment with other published operation sets (op set), use cases that require multiple new operations, compatibility with implementations, etc.
- **operation specific** - _PROPOSED_ - issues regarding the specification of a single operation or small number of operations
```

Are we making someone unhappy if we standardize on _operation_? :)

I note the spec needs some regexp action for this because we use operation and operator interchangeably.

Author: It looks like most frameworks (TF, ONNX, PyTorch) use "operator" more frequently in this context, e.g. "operator set", "operators", "operator compatibility", "operator coverage".  (thanks @a-sully) 

_Operators_ perform _operations_, so the phrase _"a platform operator for the operation"_ correctly uses both. 

For the labels, I think "operator" would align with more common usage. But I'm still new to all this!

Reviewer: Let’s go with the most popular name for the labels. We can adjust these later as needed.
